### PR TITLE
jax.numpy indexing: lower to dynamic_slice for more cases

### PR DESCRIPTION
**Pull request by jakevdp**
Built on #15390 in order for checkify tests to continue to pass

Replaces #15347, which used static slices.

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3936,39 +3938,106 @@ def replace(tup, val):
...
+
+  simple_revs = {i for i, ind in enumerate(idx) if _is_simple_reverse_slice(ind)}
+  int_indices = {i for i, (ind, size) in enumerate(zip(idx, arr.shape))
+                 if _is_valid_integer_index_for_slice(ind, size, mode)}
+  contiguous_slices = {i for i, ind in enumerate(idx) if _is_contiguous_slice(ind)}
+
+  # For sharded inputs, partial slices lead to incorrect sharding semantics, so
+  # we fall back to gather
+  # TODO(yashkatariya): fix dynamic_slice with sharding
+  is_sharded = isinstance(arr, ArrayImpl) and not dispatch.is_single_device_sharding(arr.sharding)
```
Comment (yashk2810) [inline jax/_src/numpy/lax_numpy.py:3988]: nit: format under 80 characters?

Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:3988]: Are we doing that now?

**Review by yashk2810 — APPROVED** [review]

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -3936,39 +3938,106 @@ def replace(tup, val):
...
+  all_empty_slices = _all(idx[i].indices(arr.shape[i]) == (0, arr.shape[i], 1)
+                          for i in contiguous_slices)
+  if is_sharded and (int_indices or not all_empty_slices):
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:3991]: The comment in line 3985 was meant to cover this... is there more information you'd like in that comment?

File: jax/_src/numpy/lax_numpy.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yashk2810) [reply jax/_src/numpy/lax_numpy.py:3992]: Just what you told me offline :)

The `all_empty_slices` is confusing. I don't know what that means.

You said: `This means that if there are integer indices (like x[0] or slices that aren't empty (like x[:2] as opposed to x[:]) then we avoid dynamic_slice and fall back to gather`.. maybe reword and add that?

Not a big deal if you want to skip!

[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:3992]: if I named the variable `all_slices_are_empty` would that clarify things?

[the remark this answers is not in the document]
Comment (yashk2810) [reply jax/_src/numpy/lax_numpy.py:3992]: Sure, but what does `empty` mean?

`arr[slice(None), slice(None), slice(None)]` is empty?

[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py]: OK, I added more context, let me know what you think

[the remark this answers is not in the document]
Comment (yashk2810) [reply jax/_src/numpy/lax_numpy.py]: Looks good! partial_slices is very clear! Thank you!
